### PR TITLE
Implement serialization for ScalarValue::FixedSizeBinary 

### LINKED ISSUE #3928: Implement serialization for `ScalarValue::FixedSizeBinary`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

@maxburke's PR added `ScalarFixedSizeBinary` support:  https://github.com/apache/arrow-datafusion/pull/3911 

However, it did not implement the serialization code to send `ScalarValue::FixedSizeBinary` over the network 

**Describe the solution you'd like**

Fill out the code in datafusion/proto/src/to_proto.rs

```rust
            scalar::ScalarValue::FixedSizeBinary(_, _) => {
                return Err(Error::General(
                    "FixedSizeBinary is not yet implemented".to_owned(),
                ))
            }
```

And add a test for round trip serialization

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by retikulum**
# Which issue does this PR close?

Closes #3928.

 # Rationale for this change

Implement serialization for `ScalarValue::FixedSizeBinary` for communicating with protobuf.

I added `ScalarFixedSizeBinary` as message in .proto file. I am not sure about if it is the correct way.

My other concern is one of the test  is `ScalarValue::FixedSizeBinary(0, None)` but what is the length of None?

### CODE REVIEW

File: datafusion/proto/proto/datafusion.proto [hunk 1]
```diff
@@ -760,6 +760,11 @@ message StructValue {
   repeated Field fields = 3;
 }
 
+message ScalarFixedSizeBinary{
```
[the remark this answers is not in the document]
Comment (retikulum) [reply datafusion/proto/proto/datafusion.proto:763]: Thanks for feedback. While I was looking for similar definitons, I came acrros to this and implement it in this way. https://github.com/apache/arrow-datafusion/blob/48f73c6af3b0cc747c38b4a9c7a610f4630e8736/datafusion/proto/proto/datafusion.proto#L711-L714

However, I am open to fix in a way to comply with preferred code style. If this is the only thing that needs to be changed, I will create a commit.

File: datafusion/proto/proto/datafusion.proto [hunk 2]
```diff
@@ -834,6 +840,7 @@ enum PrimitiveScalarType{
 
     BINARY = 25;
     LARGE_BINARY = 26;
+    FIXED_SIZE_BINARY = 29;
```
Comment (alamb) [inline datafusion/proto/proto/datafusion.proto:843]: I removed the duplication with PrimitiveScalarType and `ScalarValue` in https://github.com/apache/arrow-datafusion/commit/54d2870a56d8d8f914a617a7e2d52e387ef5dba2 so when you are resolving the merge conflict you should be able to just remove this code

File: datafusion/proto/src/lib.rs [hunk 3]
```diff
@@ -556,6 +556,11 @@ mod roundtrip_tests {
                     Field::new("a", DataType::Boolean, false),
                 ]),
             ),
+            ScalarValue::FixedSizeBinary(
+                b"bar".to_vec().len() as i32,
+                Some(b"bar".to_vec()),
+            ),
+            ScalarValue::FixedSizeBinary(0, None),
```
Comment (alamb) [inline datafusion/proto/src/lib.rs:563]: I suggest adding a None case for non zero length as well

```suggestion
            ScalarValue::FixedSizeBinary(0, None),
            ScalarValue::FixedSizeBinary(5, None),
```

File: datafusion/proto/src/to_proto.rs [hunk 4]
```diff
@@ -1064,10 +1064,13 @@ impl TryFrom<&ScalarValue> for protobuf::ScalarValue {
                     Value::LargeBinaryValue(s.to_owned())
                 })
             }
-            scalar::ScalarValue::FixedSizeBinary(_, _) => {
-                return Err(Error::General(
-                    "FixedSizeBinary is not yet implemented".to_owned(),
-                ));
+            scalar::ScalarValue::FixedSizeBinary(length, val) => {
```
Comment (alamb) [inline datafusion/proto/src/to_proto.rs:1067]: ❤️

**Review by alamb — COMMENTED** [review]
Thanks @retikulum and @isidentical 
I think this PR is looking good -- it just needs a rebase or merge and resolve the conflicts and it should be good to go

**Comment by retikulum** [comment]
Thanks @alamb for review and sorry for lots of commits. These conflicts are great for improving my git/rust skills :)

**Comment by alamb** [comment]
> Thanks @alamb for review and sorry for lots of commits. These conflicts are great for improving my git/rust skills :)

Thank you for putting up with it. It takes some getting used to but then it is quite cool

**Pull request merged**